### PR TITLE
Avoid using OpenThread for out of process SetThreadContext debugging

### LINKED ISSUE #107263: After upgrading to .NET 9, simple C# statement triggers fatal error in Visual Studio 2022 when Control-C is used to exit
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/After-upgrading-to-NET-9-simple-C-sta/10728272)._

---
[severity:I'm unable to use this version]
A complete description of the problem, including screenshots and code that triggers the error can be found here:

https://stackoverflow.com/questions/78890260/after-upgrading-to-net-9-simple-c-sharp-statement-triggers-fatal-error-in-visu

---
#### Feedback Bot on 8/22/2024, 02:14 AM: 

(private comment, text removed)
#### Tom McDonald [MSFT] on 8/23/2024, 04:20 PM: 

(private comment, text removed)
#### David Sworder on 8/23/2024, 04:28 PM: 

(private comment, text removed)
#### Tom McDonald [MSFT] on 8/26/2024, 10:28 PM: 

(private comment, text removed)
#### David Sworder on 8/27/2024, 05:35 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by tommcdon**
In .NET 9, CET (Control-flow Enforcement Technology) has been enabled by default (see https://github.com/dotnet/runtime/pull/103007).  The debugger has special handling when is CET enabled for breakpoints, stepping, and func-eval (see https://github.com/dotnet/runtime/pull/73572).  A customer reported regression in .NET 9 related to pressing "control-c".  Internally, when a managed control-c handler function is invoked, we try to set thread context out of process and fail to obtain the target thread's handle.  We found that once Control-C is pressed, Windows seems to change the access permissions to the process, causing attempts to retrieve the target thread's handle via to OpenThread to fail (the same happens with a C++ repro).  This PR updates the out-of-proc setthreadcontext handling logic to retrieve the handle using the DAC to workaround the access denied in OpenThread. 

Fixes #107263

### CODE REVIEW

**Comment by hoyosjs** [comment]
If this is getting blocked in Windows, could it be that such operations are now unstable?

**Comment by tommcdon** [comment]
> If this is getting blocked in Windows, could it be that such operations are now unstable?

In my testing, only OpenThread fails.  All other operations seem to work as expected.

File: src/coreclr/debug/di/process.cpp [hunk 1]
```diff
@@ -11157,17 +11157,49 @@ void CordbProcess::HandleSetThreadContextNeeded(DWORD dwThreadId)
...
-        FALSE, // thread handle is not inheritable.
-        dwThreadId);
+    // Before we can read the left side context information, we must:
+    // 1. obtain the thread handle
+    // 2. suspened the thread
+    // 3. read the thread context, and from that read the pointer to the left-side context and the size of the context
+    // 4. then we can perform the actual SetThreadContext operation
+    // 5. lastly, we must resume the thread
+    // For the first step of obtaining the thread handle, 
+    // we have previously attempted to use ::OpenThead to get a handle to the thread.
```
Comment (mikelle-rogers) [inline src/coreclr/debug/di/process.cpp:11167]:
```suggestion
    // we have previously attempted to use ::OpenThread to get a handle to the thread.
```

**Review by mikelle-rogers — COMMENTED** [review]
Code looks good to me.

File: src/coreclr/debug/di/process.cpp [hunk 2]
```diff
@@ -11157,17 +11157,49 @@ void CordbProcess::HandleSetThreadContextNeeded(DWORD dwThreadId)
...
+    if (hOutOfProcThread == NULL)
     {
-        LOG((LF_CORDB, LL_INFO10000, "RS HandleSetThreadContextNeeded - Unexpected result from OpenThread\n"));
+        LOG((LF_CORDB, LL_INFO10000, "RS HandleSetThreadContextNeeded - Failed to get thread handle\n"));
         ThrowHR(E_UNEXPECTED);
     }
 
+    // Duplicate the thread handle to the current process
+    HandleHolder hThread;
+    BOOL fSuccess = DuplicateHandle(UnsafeGetProcessHandle(), hOutOfProcThread, ::GetCurrentProcess(), &hThread, 0, FALSE, DUPLICATE_SAME_ACCESS);
```
[the remark this answers is not in the document]
Comment (tommcdon) [reply src/coreclr/debug/di/process.cpp:11195]: It differentiates the global namespace from the current class.  We are explicitly calling kernel32!GetCurrentProcess.

File: src/coreclr/debug/di/process.cpp [hunk 3]
```diff
@@ -11157,17 +11157,49 @@ void CordbProcess::HandleSetThreadContextNeeded(DWORD dwThreadId)
...
-    if (hThread == NULL)
+    HANDLE hOutOfProcThread = pDAC->GetThreadHandle(pThread->m_vmThreadToken);
+    if (hOutOfProcThread == NULL)
     {
-        LOG((LF_CORDB, LL_INFO10000, "RS HandleSetThreadContextNeeded - Unexpected result from OpenThread\n"));
+        LOG((LF_CORDB, LL_INFO10000, "RS HandleSetThreadContextNeeded - Failed to get thread handle\n"));
         ThrowHR(E_UNEXPECTED);
     }
 
+    // Duplicate the thread handle to the current process
```
[the remark this answers is not in the document]
Comment (tommcdon) [reply src/coreclr/debug/di/process.cpp:11193]: Correct - it copies the thread handle from the left side to the right side

**Review by noahfalk — APPROVED** [review]

**Pull request merged**